python/rmm/rmm.py

```diff
@@ -233,6 +138,78 @@ def get_info(stream=0):
     return librmm.rmm_getinfo(stream)
 
 
+try:
+    import numba
+except Exception:
+    numba = None
+    _numba_memory_manager = None
+
+if numba:
+    class RMMNumbaManager(HostOnlyCUDAMemoryManager):
+        def __init__(self, logging=False):
+            super().__init__()
+            self._initialized = False
+            self._logging = logging
+
+        def memalloc(self, nbytes, stream=0):
+            buf = librmm.DeviceBuffer(size=nbytes, stream=stream)
+            ctx = cuda.current_context()
+            ptr = ctypes.c_uint64(int(buf.ptr))
+            mem = MemoryPointer(ctx, ptr, nbytes, owner=buf)
+            return mem
+
+        def get_ipc_handle(self, memory, stream=0):
+            """
+            Get an IPC handle from the DeviceArray ary with offset modified by
+            the RMM memory pool.
+            """
+            # Not a very clean implementation - may want to implement something
+            # at the C++ layer for this, and also not rely on borrowing bits of
+            # Numba internals to initialise ipchandle.
+            ipchandle = (ctypes.c_byte * 64)()  # IPC handle is 64 bytes
+            cuda.cudadrv.memory.driver_funcs.cuIpcGetMemHandle(
+                ctypes.byref(ipchandle),
+                memory.owner.handle,
+            )
+            source_info = cuda.current_context().device.get_device_identity()
+            ptr = memory.device_ctypes_pointer.value
+            offset = librmm.rmm_getallocationoffset(ptr, stream)
```

Comment:
@harrism I forgot about `rmmGetAllocationOffset`. What do you think the future of this API is given the changes we're proposing in https://github.com/rapidsai/rmm/issues/301 https://github.com/rapidsai/rmm/issues/302 https://github.com/rapidsai/rmm/issues/305.

Looking at the implementation, it's independent of any resource since it uses the [CUDA driver API directly](https://github.com/rapidsai/rmm/blob/branch-0.13/src/rmm.cpp#L99). So we don't have to worry about burdening resources with implementing this functionality.

Reply:
I'm using the CUDA driver here because it was a route I could see to get this to work - I found that some of the Numba tests didn't work with the existing implementation from the deleted `get_ipc_handle` above (I can't recall the exact details unfortunately). It feels as though it would be cleaner from this side if resources did implement this functionality - I don't feel confident this implementation will work robustly with all possible resource implementations, though I'm not familiar enough with them to state specifically what issues could arise.

Reply:
> I'm using the CUDA driver here because it was a route I could see to get this to work 

To be clear, I was referring to the use of the driver API in the implementation of `rmmGetAllocationOffset` which uses `cuMemGetAddressRange`.

This is necessary because [`cudaIpcGetMemHandle`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g8a37f7dfafaca652391d0758b3667539) requires the _base_ pointer. When using a pool resource, a given pointer `p` resides somewhere within a larger `cudaMalloc'd` chunk of memory pointed to by `b`. In order to get an IPC handle for `p` we need to get `b` to pass into `cudaIpcGetMemHandle`.

Reply:
@jrhemstad yes I had thought about this yesterday when I first saw this PR. I think when we  remove `rmmInitialize`, `rmmAlloc` and friends  that `rmmGetAllocationOffset` should change to the following free function.

`ptrdiff_t  rmm::allocation_offset(void* ptr);`

Note no need for a stream (currently unused), and throws exceptions rather than returns errors. Python code would use this rather than calling the CUDA driver API directly.

Reply:
Makes sense. Would RMM still be able to be header only in that case? Otherwise the driver lib will need to be linked in to the user app to support the `cuMemGetAddressRange` call.

Reply:
Ugh. That's true.

Reply:
Darn only driver API function we use! I guess that's not that bad... and we may choose to use some of the new virtual memory management APIs in the future, which are also driver only...